doc/source/dev/missing-bits.rst

```diff
@@ -0,0 +1,235 @@
+or NumPy repositories.  If a style issue is important enough that a reviewer
+will require a change before merging, then it is important enough to be
+documented--at least for cases where the issue can be resolved with a simple
+rule.
+
+
+Coding Style and Guidelines
+---------------------------
+
+Required keyword names
+~~~~~~~~~~~~~~~~~~~~~~
+For new functions or methods with more than a few arguments, all parameters
+after the first few "obvious" ones should *require* the use of the keyword
+when given.  This is implemented by including ``*`` at the appropriate point
+in the signature.
+
+For example, a function ``foo`` that operates on a single array but that has
+several optional parameters (say ``method``, ``flag``, ``rtol`` and ``atol``)
+would be defined as::
+
+    def foo(x, *, method='basic', flag=False, rtol=1.5e-8, atol=1-12):
+        ...
+
+To call ``foo``, all parameters other than ``x`` must be given with an
+explicit keyword, e.g. ``foo(arr, rtol=1e-12, method='better')``.
+
+This forces callers to give explicit keyword parameters (which most users
+would probably do anyway even without the use of ``*``), *and* it means
+additional parameters can be added to the function anywhere after the
+``*``; new parameters do not have to be added after the existing parameters.
+
+
+Test functions from `numpy.testing`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In new code, don't use `assert_almost_equal`, `assert_approx_equal` or
+`assert_array_almost_equal`. This is from the docstrings of these
+functions::
+
+    It is recommended to use one of `assert_allclose`,
+    `assert_array_almost_equal_nulp` or `assert_array_max_ulp`
+    instead of this function for more consistent floating point
+    comparisons.
+
+For more information about writing unit tests, see `Testing Guidelines
+<https://docs.scipy.org/doc/numpy/reference/testing.html>`_.
+
+Testing that expected exceptions are raised
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When writing a new test that a function call raises a certain exception,
+the preferred style is to use ``pytest.raises`` as a context manager, with
+the code that is supposed to raise the exception in the code block defined
+by the context manager.  The `match` keyword argument of ``pytest.raises``
+is given with enough of the expected error message attached to the exception
+to ensure that the expected exception is raised.
+
+For example, the function `scipy.stats.zmap` is supposed to raise a
+``ValueError`` if the input contains ``nan`` and ``nan_policy`` is ``"raise"``.
+A test for this is::
+
+    scores = np.array([1, 2, 3])
+    compare = np.array([-8, -3, 2, 7, 12, np.nan])
+    with pytest.raises(ValueError, match='input contains nan'):
+        stats.zmap(scores, compare, nan_policy='raise')
+
+The ``match`` argument ensures that the test doesn't pass by raising
+a ``ValueError`` that is not related to the input containing ``nan``.
```

Comment:
I would not recommend any use or `re` nor `capsys`. You should be able to do this use raw strings `match=r"...."`

Reply:
Use of `re` is a separate question, though.

I'm asking whether the f-string substitution should be matched to prevent issues like that one.

Reply:
That's an [interesting case](https://github.com/scipy/scipy/pull/15566/files/21f7971ddf0b611a3c802dc837cadbe7531681b0#diff-49422b768d9abc96ae02952d0fe0a68832b6ce211857185449ebb104cbe02681); the final fix that got merged was very simple, but there was no regression test added that would ensure that the fix worked.   That particular case is simple enough that it doesn't really bother me, and I don't know if similar situations are common enough to justify creating a guideline.  So all I can offer here is a shrug, and suggest we not attempt to impose written guidelines until we have more cases.

I also don't feel like I have anything to say about using pytest `capsys` at the moment.  I expect this "missing bits" document to evolve, so if patterns emerge that warrant additional guidelines for developers, we can add them later.

I'm leaving this conversation open, in case anyone else wants to chime in.